### PR TITLE
Add external camera source to VisionManager.create.

### PR DESCRIPTION
**Pull request by yunikkk**
Also, added playback of media files through MediaCodec. Can be used to substitute video source with videos.

### CODE REVIEW

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt [hunk 1]
```diff
@@ -184,7 +174,7 @@ object VisionManager : ARDataProvider {
      * You should [destroy] when Vision SDK is no longer needed to release all resources.
      * No-op if called while SDK is created already.
      */
-    fun create() {
+    fun create(videoSource: VideoSource = CameraVideoSourceImpl(application)) {
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: External video source is passed here. Default is still good old camera2 implementation.

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt [hunk 2]
```diff
@@ -297,7 +288,6 @@ object VisionManager : ARDataProvider {
             return
         }
 
-        videoSource.release()
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: `release` method removed, it content moved to `detach`

File: MapboxVision/src/main/java/com/mapbox/vision/camera/CameraParamsListener.kt [hunk 3]
```diff
@@ -1,8 +0,0 @@
-package com.mapbox.vision.camera
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/camera/CameraParamsListener.kt]: No more listener, camera manager sets parameters itself.

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/VideoSource.kt [hunk 4]
```diff
@@ -1,22 +1,15 @@
...
 
-    fun isAttached(): Boolean
-
     fun attach(videoSourceListener: VideoSourceListener)
 
     fun detach()
 
-    fun release()
-
     fun stopVideoRecording()
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/VideoSource.kt]: Video recording methods still in progress of moving to separate entities.

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/CameraVideoSourceImpl.kt [hunk 5]
```diff
@@ -99,25 +72,6 @@ internal class CameraVideoSourceImpl(
         }
     }
 
-    /** A [CameraCaptureSession.CaptureCallback] that handles events related to capture.  */
-    private val captureCallback = object : CameraCaptureSession.CaptureCallback() {
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/CameraVideoSourceImpl.kt]: Callback can be replaced with `null` in `setRepeatingRequest`

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/CameraVideoSourceImpl.kt [hunk 6]
```diff
@@ -183,40 +138,12 @@ internal class CameraVideoSourceImpl(
...
-            yuv2RgbRender.useBitmap(useBitmap)
-        }
-
-        this.useBitmap = useBitmap
-    }
-    // end
-
-
-    // Render actions
-    override fun getRgbBytesArray(): ByteArray = rgbaBytes
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/CameraVideoSourceImpl.kt]: Simplified a bit here, now `yuv2rgb` owns this byte array and notifies us.

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/CameraVideoSourceImpl.kt [hunk 7]
```diff
@@ -328,96 +249,78 @@ internal class CameraVideoSourceImpl(
...
-                                    videoSourceListener?.onNewCameraParams(cameraParamsData)
-                                }
-                            }
-                        }
+                val physicalSensorSize = characteristics.get(CameraCharacteristics.SENSOR_INFO_PHYSICAL_SIZE)!!
+                val focalInPixelsX: Float = focalLength * previewSize.width / physicalSensorSize.width
+                val focalInPixelsY: Float = focalLength * previewSize.height / physicalSensorSize.height
 
-                this.cameraId = cameraId
+                cameraParams = CameraParamsData(
```
Comment (yunikkk) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/CameraVideoSourceImpl.kt]: Parameters are now counted here once and updated when listener attaches in `attach`

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt [hunk 8]
```diff
@@ -492,10 +505,11 @@ object VisionManager : ARDataProvider {
     }
 
     private fun requestCoreUpdate() {
-        val lastCoreUpdateStartTime = System.currentTimeMillis()
+        coreUpdateThreadHandler.postDelayed(
+                { requestCoreUpdate() },
+                CORE_UPDATE_DELAY_MILLIS
```
Comment (BenDenen) [inline MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: As far as I remember we should update the core every 30 milliseconds minus time for method execution?

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/YuvAllocation2Rgb.kt [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import android.graphics.ImageFormat
+import android.renderscript.Allocation
+import android.renderscript.Element
+import android.renderscript.RenderScript
+import android.renderscript.ScriptIntrinsicYuvToRGB
+import android.renderscript.Type
+import android.util.Size
+import android.view.Surface
+
+internal class YuvAllocation2Rgb(
```
Comment (BenDenen) [inline MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/YuvAllocation2Rgb.kt]: What the reason to  rename Yuv2RgbRender to YuvAllocation2Rgb?

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt [hunk 10]
```diff
@@ -194,24 +190,43 @@ object VisionManager : ARDataProvider {
...
+                val videoRecorder = SurfaceVideoRecorder.MediaCodecPersistentSurfaceImpl(
+                        application = VisionManager.application,
+                        buffersDir = FileUtils.getDataDirPath(VisionManager.application),
+                        recordingListener = recordingListener,
+                        sensorOrientation = videoSource.sensorOrientation,
+                        frameWidth = videoSource.getSourceWidth(),
+                        frameHeight = videoSource.getSourceHeight()
+                )
+                this.videoRecorder = videoRecorder
+                videoSource.setRecordingSurface(videoRecorder.surface)
```
[the remark this answers is not in the document]
Comment (yunikkk) [reply MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: Still in progress, discussed it.

File: MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt (hunk 8 above)
Comment (yunikkk) [reply MapboxVision/src/main/java/com/mapbox/vision/VisionManager.kt]: It posts the next update for 30 FPS and then requests update, so the FPS is the actual 30 FPS

File: MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/YuvAllocation2Rgb.kt (hunk 9 above)
Comment (yunikkk) [reply MapboxVision/src/main/java/com/mapbox/vision/video/videosource/camera/YuvAllocation2Rgb.kt]: Well, it doesn't render anything, but instead provides Allocation with Surface, then converts frames to Rgb.
WDYT?

**Pull request closed without merging**